homeassistant/components/sql/sensor.py

```diff
@@ -115,73 +118,62 @@ def setup_platform(
-
-    @property
-    def name(self):
-        """Return the name of the query."""
-        return self._name
-
-    @property
-    def native_value(self):
-        """Return the query's current state."""
-        return self._state
-
-    @property
-    def native_unit_of_measurement(self):
-        """Return the unit of measurement."""
-        return self._unit_of_measurement
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        return self._attributes
-
-    def update(self):
+        self._attr_unique_id = f"{db_url}, {query}, {column}, {value_template}"
+        self._attr_extra_state_attributes = {}
+
+    def update(self) -> None:
         """Retrieve sensor data from the query."""
 
         data = None
+        self._attr_extra_state_attributes = {}
+        sess: scoped_session = self.sessionmaker()
         try:
-            sess = self.sessionmaker()
             result = sess.execute(self._query)
-            self._attributes = {}
-
-            if not result.returns_rows or result.rowcount == 0:
-                _LOGGER.warning("%s returned no results", self._query)
-                self._state = None
-                return
-
-            for res in result.mappings():
-                _LOGGER.debug("result = %s", res.items())
-                data = res[self._column_name]
-                for key, value in res.items():
-                    if isinstance(value, decimal.Decimal):
-                        value = float(value)
-                    if isinstance(value, datetime.date):
-                        value = str(value)
-                    self._attributes[key] = value
         except sqlalchemy.exc.SQLAlchemyError as err:
             _LOGGER.error(
                 "Error executing query %s: %s",
                 self._query,
                 redact_credentials(str(err)),
             )
             return
-        finally:
-            sess.close()
+
+        _LOGGER.debug("Result %s, ResultMapping %s", result, result.mappings())
+
+        for res in result.mappings():
+            _LOGGER.debug("result = %s", res.items())
+            data = res[self._column_name]
+            for key, value in res.items():
+                if isinstance(value, float):
+                    value = float(value)
+                if isinstance(value, date):
+                    value = value.isoformat()
+                self._attr_extra_state_attributes[key] = value
 
         if data is not None and self._template is not None:
-            self._state = self._template.async_render_with_possible_json_value(
-                data, None
+            self._attr_native_value = (
+                self._template.async_render_with_possible_json_value(data, None)
             )
         else:
-            self._state = data
+            self._attr_native_value = data
+
+        if not data:
+            _LOGGER.warning("%s returned no results", self._query)
+
+        sess.close()
```

Comment:
not sure about what happens to `sess` is exception if previously thrown (that's the reason before for the `finally`)

Reply:
I'm sure I tested but let me test again just for precaution

Reply:
Works fine